<a href="https://colab.research.google.com/github/mohfujulm/AI_Assignment4/blob/main/Judging%20Flowers.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [146]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

In [147]:
# Load the dataset (load remotely, not locally)
url = "https://gist.githubusercontent.com/netj/8836201/raw/6f9306ad21398ea43cba4f7d537619d0e07d5ae3/iris.csv"
iris = pd.read_csv(url)

#0 = Setosa, 1 = Versicolor, 2 = Virginica
iris.variety.loc[iris.variety == 'Setosa'] = 0
iris.variety.loc[iris.variety == 'Versicolor'] = 1
iris.variety.loc[iris.variety == 'Virginica'] = 2
irisNames = ["Setosa","Versicolor","Virginica"]

# Output the first 15 rows of the data
iris.head(15)

#warning indicating we are working on a copy, but does not matter for the purposes of this assignment

C:\Users\MJ\anaconda3\lib\site-packages\pandas\core\indexing.py:1732: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


sepal.length  sepal.width  petal.length  petal.width variety
0            5.1          3.5           1.4          0.2       0
1            4.9          3.0           1.4          0.2       0
2            4.7          3.2           1.3          0.2       0
3            4.6          3.1           1.5          0.2       0
4            5.0          3.6           1.4          0.2       0
5            5.4          3.9           1.7          0.4       0
6            4.6          3.4           1.4          0.3       0
7            5.0          3.4           1.5          0.2       0
8            4.4          2.9           1.4          0.2       0
9            4.9          3.1           1.5          0.1       0
10           5.4          3.7           1.5          0.2       0
11           4.8          3.4           1.6          0.2       0
12           4.8          3.0           1.4          0.1       0
13           4.3          3.0           1.1          0.1       0
14           5.8          4.0           1.2          0.2       0

In [159]:
# Display a summary of the table information (number of datapoints, etc.)

iris.info()

numberDatapoints = iris["variety"].count()
averageSepalLength = iris["sepal.length"].mean()
averageSepalWidth = iris["sepal.width"].mean()
averagePetalLength = iris["petal.length"].mean()
averagePetalWidth = iris["petal.width"].mean()

print("\nNumber of datapoints: ", round(numberDatapoints, 3))
print("Average sepal length: ", round(averageSepalLength, 3))
print("Average sepal width: ", round(averageSepalWidth, 3))
print("Average petal length: ", round(averagePetalLength, 3))
print("Average petal width: ", round(averagePetalWidth, 3))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal.length  150 non-null    float64
 1   sepal.width   150 non-null    float64
 2   petal.length  150 non-null    float64
 3   petal.width   150 non-null    float64
 4   variety       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB

Number of datapoints:  150
Average sepal length:  5.843
Average sepal width:  3.057
Average petal length:  3.758
Average petal width:  1.199


### About the dataset

The iris dataset contains information about some properties of three species of Iris flowers (Iris setosa, Iris virginica, and Iris versicolor).  Namely, these properties are the lengths and widths of both the sepals and petals of the flower.  The final column of the dataset contains the variety of flower corresponding to each datapoint containing these four properties.  

As for features and labels, our features (x) are the four given properties of the Iris flowers - sepal and petal length/width.  Our label (y) is the variety of the flower these correspond to, represented by the discrete set [0, 1, 2], with 0 mapping to the setosa flower, 1 mapping to the virginica flower, and 2 mapping to the versicolor flower.

In [160]:
# Take the dataset and split it into our features (X) and label (y)
irisX = iris[["sepal.length", "sepal.width", "petal.length", "petal.width"]]
irisY = iris["variety"]
irisY = irisY.astype('int')

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
irisX_train, irisX_test, irisY_train, irisY_test = train_test_split(irisX, irisY, train_size = 0.9, test_size=0.1, random_state = 150)


## Logistic Regression

In [161]:
# i. Use sklearn to train a LogisticRegression model on the training set

reg = linear_model.LogisticRegression(solver='lbfgs', max_iter=1000)
reg.fit(irisX_train, irisY_train)

# ii. For a sample datapoint, predict the probabilities for each possible class

# Sample datapoint from test set 
sampleTestX = irisX_test.tail(1).values
sampleTestY = irisY_test.tail(1).values

print("Sample datapoint: ")
print("    Sepal length: ", sampleTestX[0][0], ", Sepal width: ", sampleTestX[0][1])
print("    Petal length: ", sampleTestX[0][2], ", Petal width: ", sampleTestX[0][3]) 
print("\nActual variety of flower: \n   ", irisNames[sampleTestY[0]])     
print("\nProbabilities of class of flower from model for this datapoint: \n    Setosa: ", round(reg.predict_proba(sampleTestX)[0][0]*100,2), "%")
print("    Versicolor: ", round(reg.predict_proba(sampleTestX)[0][1] * 100,2), "%")
print("    Virginica: ", round(reg.predict_proba(sampleTestX)[0][2] * 100,2), "%")

#print(reg.predict(sampleTestX))

# iii. Report on the score for Logistic regression model, what does the score measure?

print("\nModel Score: \n", reg.score(irisX_test, irisY_test))

# iv. Extract the coefficents and intercepts for the boundary line(s)
print("\nCoefficients: \n", reg.coef_)
print("\nIntercepts: \n", reg.intercept_)

Sample datapoint: 
    Sepal length:  6.7 , Sepal width:  3.3
    Petal length:  5.7 , Petal width:  2.5

Actual variety of flower: 
    Virginica

Probabilities of class of flower from model for this datapoint: 
    Setosa:  0.0 %
    Versicolor:  1.37 %
    Virginica:  98.63 %

Model Score: 
 0.9333333333333333

Coefficients: 
 [[-0.42110328  0.95919094 -2.47288555 -1.06204802]
 [ 0.58933891 -0.24325485 -0.26514129 -0.86002417]
 [-0.16823563 -0.71593609  2.73802684  1.92207219]]

Intercepts: 
 [  9.68983839   1.73650518 -11.42634357]


Above we see the score for the Logistic Regression model is about 0.933, a metric calculated using our test values that represents the mean accuracy of our model.  It measures the number of correct predictions made by our model divided by the total number of all predictions, with perfect accuracy being a score of 1.  Thus, a score of 0.933 indicates that the model is overall highly accurate at determining the feature of a particular iris flower given one of its four features using the logistic regression model.

The extracted intercepts and coefficients of our equation are listed above.  

## Support Vector Machine

In [165]:
# i. Use sklearn to train a Support Vector Classifier on the training set

svmClf = svm.SVC(probability = True)
svmClf.fit(irisX_train, irisY_train)

# ii. For a sample datapoint, predict the probabilities for each possible class

#Utilizing sample datapoint from before
print("Sample datapoint: ")
print("    Sepal length: ", sampleTestX[0][0], ", Sepal width: ", sampleTestX[0][1])
print("    Petal length: ", sampleTestX[0][2], ", Petal width: ", sampleTestX[0][3]) 
print("\nActual variety of flower: \n   ", irisNames[sampleTestY[0]])  

print("\nProbabilities of class of flower from model for this datapoint: \n    Setosa: ", round(svmClf.predict_proba(sampleTestX)[0][0]*100,2), "%")
print("    Versicolor: ", round(svmClf.predict_proba(sampleTestX)[0][1] * 100,2), "%")
print("    Virginica: ", round(svmClf.predict_proba(sampleTestX)[0][2] * 100,2), "%")

#print(svmReg.predict(sampleTestX))

# iii. Report on the score for the SVM, what does the score measure?

print("\nModel Score: \n", svmReg.score(irisX_test, irisY_test))

Sample datapoint: 
    Sepal length:  6.7 , Sepal width:  3.3
    Petal length:  5.7 , Petal width:  2.5

Actual variety of flower: 
    Virginica

Probabilities of class of flower from model for this datapoint: 
    Setosa:  0.77 %
    Versicolor:  0.14 %
    Virginica:  99.1 %

Model Score: 
 0.9333333333333333


Above, the score for the SVM model was about 0.933 for this shuffle of the data, once again indicating that the model was highly accurate in predicting the type of Iris flower.  The score is the same as for the Logistic Regression model, meaning that the two models performed about the same in terms of overall accuracy.  For the sample datapoint (using the same for both models), the SVM model seemed to output a marginally higher probability score for the correct flower type, by about 0.4%.

## Neural Network

In [166]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set

MLPclf = MLPClassifier(activation = 'logistic', solver='lbfgs', max_iter=2000)
MLPclf.fit(irisX_train, irisY_train)

# ii. For a sample datapoint, predict the probabilities for each possible class

#Utilizing sample datapoint from before
print("Sample datapoint: ")
print("    Sepal length: ", sampleTestX[0][0], ", Sepal width: ", sampleTestX[0][1])
print("    Petal length: ", sampleTestX[0][2], ", Petal width: ", sampleTestX[0][3]) 
print("\nActual variety of flower: \n   ", irisNames[sampleTestY[0]])  

print("\nProbabilities of class of flower from model for this datapoint: \n    Setosa: ", round(MLPclf.predict_proba(sampleTestX)[0][0]*100,2), "%")
print("    Versicolor: ", round(MLPclf.predict_proba(sampleTestX)[0][1] * 100,2), "%")
print("    Virginica: ", round(MLPclf.predict_proba(sampleTestX)[0][2] * 100,2), "%")

#print(MLPreg.predict(sampleTestX))

# iii. Report on the score for the Neural Network, what does the score measure?

print("\nModel Score Test Data: \n", MLPclf.score(irisX_test, irisY_test))

# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)

Sample datapoint: 
    Sepal length:  6.7 , Sepal width:  3.3
    Petal length:  5.7 , Petal width:  2.5

Actual variety of flower: 
    Virginica

Probabilities of class of flower from model for this datapoint: 
    Setosa:  0.0 %
    Versicolor:  0.0 %
    Virginica:  100.0 %

Model Score Test Data: 
 0.9333333333333333


The score for the neural network was once again 0.933, identical to the performance of the preceding two models, and indicating that this model is also highly accurate in predicting the type of Iris flower given a datapoint.  Compared to the other two models, the outputted probability of the correct class of flower was at a clear 100% for this shuffle of the data, as opposed to ~ 98.6% - 99.1% in the Logistic Regression/SVM models.

The best performance I achieved was using the logistic sigmoid function for the activation function and the lbfgs solver for weight optimization.  With some changes to other parameters, I obtained similar overall accuracy scores, but the probabilities for the 3 classes of flowers were less clear cut for the sample datapoint.  With some parameters such as using the sgd solver and inverse scaling for learning rate, the score declined to values as low as 0.333.  

## K-Nearest Neighbors

In [167]:
# i. Use sklearn to 'train' a k-Neighbors Classifier

kNNclf = KNeighborsClassifier(n_neighbors = 10)
kNNclf.fit(irisX_train, irisY_train)


# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

# ii. For a sample datapoint, predict the probabilities for each possible class

#Utilizing sample datapoint from before
print("Sample datapoint: ")
print("    Sepal length: ", sampleTestX[0][0], ", Sepal width: ", sampleTestX[0][1])
print("    Petal length: ", sampleTestX[0][2], ", Petal width: ", sampleTestX[0][3]) 
print("\nActual variety of flower: \n   ", irisNames[sampleTestY[0]])  

print("\nProbabilities of class of flower from model for this datapoint: \n    Setosa: ", round(kNNclf.predict_proba(sampleTestX)[0][0]*100,2), "%")
print("    Versicolor: ", round(kNNclf.predict_proba(sampleTestX)[0][1] * 100,2), "%")
print("    Virginica: ", round(kNNclf.predict_proba(sampleTestX)[0][2] * 100,2), "%")

print("\nModel Score: \n", MLPclf.score(irisX_test, irisY_test))

# iii. Report on the score for kNN, what does the score measure?

Sample datapoint: 
    Sepal length:  6.7 , Sepal width:  3.3
    Petal length:  5.7 , Petal width:  2.5

Actual variety of flower: 
    Virginica

Probabilities of class of flower from model for this datapoint: 
    Setosa:  0.0 %
    Versicolor:  0.0 %
    Virginica:  100.0 %

Model Score: 
 0.9333333333333333


Once again, the score for the k-Neighbors Classifier was 0.933, indicating a very high accuracy equivalent to that of the previous models for this random shuffle of the Iris dataset. However, it's probability split for the sample datapoint was equivalent to only that of the MLP classifier/neural net model, with a probability of 100% for the correct flower type for this datapoint.  

## Conclusions

Overall, all of the models were very good at predicting the type of Iris flower given a datapoint of a flower's sepal length/ width and petal length/width.  They scored highly on their accuracy scores, all of them hitting 0.933 on this particular shuffle of the data.  They vary slightly depending on the particular split of data, but generally score in that range with some slight variation.  

The best performing models, given that the scores were all the same, seemed to be the K-Nearest Neighbors and Neural Net/MLP classifier models, as the predicted probability of the correct flower type was usually higher than the other models, and was the highest in this iteration of the notebook, hitting 100% probability for the actual flower type in both models, compared to 98-99% for the other models.  Even so, this difference was very slight.  It is possible that the Neural Network model performed best, compared to models like the logisitic regression because it is capable of producing more complex decision boundaries, which may be needed for this dataset where there is significant overlap between some pairs of the flower classes for certain features.  What surprised me about the exercise was that there was not much variation between the accuracy of the different models given their different approaches to classification of this dataset, and was expecting to see much more clear distinctions in scores/performance of the models.  

